doc/dev/kdb-operations.md

```diff
@@ -0,0 +1,159 @@
+
+In simpler terms, this means to the caller it looks as if `kdbGet()` had removed all keys below `parentKey` from `ks` and then loaded the data from the backends.
+Below `parentKey` the KeySet `ks` correctly represents the state of the KDB.
+For the rest of `ks` there are no such guarantees.
+
+TODO: cache correct? notifications? (just call at the end?)
+
+The basic flow of this operation is:
+
+1. Determine the backends needed to read all keys below `parentKey`
+2. Run the `init` phase on all the backends that haven't been initialized
+3. Run the `resolver` phase on all backends
+4. From now on ignore all backends, which indicated that there is no update.
+5. If all backends are now ignored, **return**.
+6. If a global cache plugin is enabled:
+   Ask the global cache plugin for the cache times for all backends.
+7. If all backends have an existing cache entry:
+   Run the `cachecheck` phase on all backends
+8. If all backends indicated the cache is still valid:
+   Ask the global cache plugin for the cached data and **return**.
+9. Run the `prestorage` and `storage` phase on all backends.
+10. Run the `poststorage` phase of all `spec:/` backends.
+11. Merge the data from all backends
+12. If enabled, run the `gopts` plugin.
+13. Run the `spec` plugin (to copy metakeys).
+14. Split data back into individual backends.
+15. Run the `poststorage` phase for all backends.
+16. Merge the data from all backends.
+17. If a global cache plugin is enabled, update cache.
+    Then **return**.
+
+> **Note:** In case of error, we abort immediately, restore `ks` to its original state and return.
+
+Influence of namespaces:
+
+- cascading and `meta:/` keys are always illegal in `ks` (should be enforced via different KeySet types)
+- `spec:/` backends only go through `init`, `resolver`, `cache`, `presetstorage` and `storage` phases as normal, but their `poststorage` phase is called earlier.
+- `dir:/`, `user:/` and `system:/` go through all phases as described above.
+- `proc:/` mountpoints go through all the phases as described above, but they are not stored in the cache.
+- `default:/` backends only go through the `poststorage` phase.
+
+## `set` Operation
+
+The purpose of the `set` operation is to write data from a `KDB` instance into backends.
+
+> **Note:** Some details of a `set` operation are defined in the [contract with backend plugins](./backend-plugins.md).
+
+Properties of `kdbSet()`:
+
+- When calling `kdbSet (kdb, ks, parentKey)` the contents (key names, values and metadata) of `ks` _will not be modified_.
+- _All keys_ in `ks` that are below `parentKey` will be persisted in the KDB, when a `kdbSet (kdb, ks, parentKey)` call returns successfully.
+- Calling `kdbSet` results in an error, if `kdbGet` wasn't called on this `KDB` instance at least once.
+
+TODO: notifications? (just call at the end?)
+
+The basic flow of this operation is:
+
+1. Check if `ks` needs sync (via `KS_FLAG_SYNC` flag), if so go to 3.
```

Comment:
Yes, I know it was deprecated, but I would still use it here.

Essentially, this is an optimization. If we set `KS_FLAG_SYNC` when a key is added to or removed from a `KeySet`, we don't need to check all keys. So if `KS_FLAG_SYNC` is set, there definitely was a change. If it is not set, there might have been a change.

>  check if previous invocations had the same number of keys.

The number of keys doesn't mean anything. I could remove key A and add key B and would still have the same number of keys.

Reply:
This would be premature optimization, or actually worse: introducing an optimization which already showed that it does not matter (and was removed because of this).

> The number of keys doesn't mean anything. I could remove key A and add key B and would still have the same number of keys.

But in this case you would have a sync flag on key B.

Reply:
> already showed that it does not matter

If you can point me to how/where this was shown, I may agree with you. However, I don't think you'll fine this evidence. At least not for the `KS_FLAG_SYNC` I want to use (the old one may have been used differently).

---

It's kind of funny. There are two issues here (`KS_FLAG_SYNC` and "number of keys") and we are arguing in the exact opposite direction of each other in both cases.

You argue the `KS_FLAG_SYNC` is premature optimization or not an optimization at all, while checking the number of keys make sense. And I argued that `KS_FLAG_SYNC` makes sense, but the number of keys check is redundant.

---

The naive approach to checking whether a `KeySet` has changed is to iterate all keys and check the `KEY_FLAG_SYNC` flags. This means any additional check _could_ be an optimization, whether this is `KS_FLAG_SYNC` or it is the size of the KeySet.

 `KS_FLAG_SYNC` if done correctly is an optimization for the case, where the `KeySet` knows it has been changed, e.g. because `ksPop` was called. 

The "number of keys" check is also an optimization for the case where the `KeySet` was changed.

Both of these do however not help in detecting unchanged `KeySet`s. `KS_FLAG_SYNC` won't be set if a simple `keySetString` was done and the number of keys will be unchanged, if I do one `ksPop` and one `ksAppendKey`.

So either both of these "optimizations" are valid, or both them "don't matter" because they are only sufficient but not necessary conditions for "KeySet changed" (i.e. `IF cond then "KeySet changed"`, but not `IF NOT cond then NOT "KeySet changed"`).

Reply:
This is a huge topic which took me a few weeks to understand. In that time I followed a few wrong ways, `KS_FLAG_SYNC` is one of them. It is probably better if we talk about it tomorrow. Maybe we can make the "backend meeting" directly after the Elektra meeting?

> If you can point me to how/where this was shown,

It is actually easy to profile for yourself (benchmarks/kdb.c): you will see which parts of `kdbGet` take (much) (CPU) time. Maybe you can do this already before the meeting (I used kcachegrind but it is good if you also check with other tools), so that we have more current data. (I did not do the profiling again with the mmap changes.)

> This means any additional check could be an optimization, whether this is KS_FLAG_SYNC or it is the size of the KeySet.

There are different levels of optimizations, the most important one is to not make useless syscalls, e.g. not to read/write a config file again which was read/written just before. To check the size of the KeySets in the individual backends is such an optimization. `KS_FLAG_SYNC` (at least in my understanding what this flag should do) alone cannot fulfill this, as you can pop a key and insert the same key again or you might add a key which changes one backend (and not others).

To have several levels of optimizations (like first checking `KS_FLAG_SYNC` and then checking per backend) is something which we should **only** do if it is clear that the code path of checking per backend is expensive. (Which it never was in my profiling. And what we actually cannot know now, as your code does not even exist yet.) The optimization is very easy to add at any time and has nothing to do with 1.0 goals. So it is simply a not well-spent time to do it now without anyone complaining.

Reply:
> It is probably better if we talk about it tomorrow. Maybe we can make the "backend meeting" directly after the Elektra meeting?

Yes let's talk about it tomorrow.

I actually think the confusion here, comes from a misunderstanding of what happens in steps 1 & 2. These steps are only meant as optimizations, to check whether `kdbSet` needs to be called at all. Maybe you thought that this already determines, which backends need to be updated, but that only happens in step 5 (details currently unspecified and not implemented).

Still for now here are some short answers to your comment:

> It is actually easy to profile for yourself [...] And what we actually cannot know now, as your code does not even exist yet.

Most of the code does exist already in this PR, but it is very much not final, so profile wouldn't make sense.

When you are profiling the results also very much depend on what code you are profiling. The code in `benchmarks/kdb.c` is very artificial. AFAICT it is also meant as a benchmark (not for profiling) of `kdbGet` and _not_ `kdbSet`.

> To check the size of the KeySets in the individual backends is such an optimization. KS_FLAG_SYNC (at least in my understanding what this flag should do) alone cannot fulfill this, as you can pop a key and insert the same key again or 

Checking the size of the _full_ `KeySet` _at this stage_ is not any better than `KS_FLAG_SYNC`. With both solutions, you can only detect when a `KeySet` has definitely been changed. However, you never cannot be sure `KeySet` is unchanged with either of these approaches.

> you might add a key which changes one backend (and not others)

This is addresses later on in step 5. This first step only determines, if we need to do a `kdbSet` at all.

I haven't yet thought about how to best do this. But probably, it will be size similar to what was done with the Split sizes. 

Note: In contrast to step 1 & 2, I don't see step 5 as an optimization. IMO it is a core feature of `kdbSet()` that it only writes to a file, if there are changes. Step 1 & 2 on the other hand are optimizations, if we remove them `kdbSet()` will behave the same, but will sometimes be slower.

> To have several levels of optimizations (like first checking KS_FLAG_SYNC and then checking per backend) is something which we should only do if it is clear that the code path of checking per backend is expensive.

This also applies to a single level of optimization. Anything the makes the code more complex without any benefit should be avoided.

---

In conclusion: I think I will write the code the way it is specified here. When I'm done, I will check whether steps 1&2 make any difference and if not, I will remove them. To me it seems like these checks should provide a benefit, but maybe you are correct and it doesn't actually matter.

Reply:
(short comments about the profiling as it is useful to do it before the meeting tomorrow)

> Most of the code does exist already in this PR, but it is very much not final, so profile wouldn't make sense.

I meant profiling the current situation (master) so that you have a feeling at which places how much time is lost and plan the optimizations based on real data.

> When you are profiling the results also very much depend on what code you are profiling. The code in benchmarks/kdb.c is very artificial.

kdbGet is also the operation where it is more important that it is fast.

> AFAICT it is also meant as a benchmark (not for profiling) of kdbGet and not kdbSet.

You can do both. It is probably also useful to note the benchmark times so that we see that it is not much slower after the PR.

Reply:
> profiling as it is useful to do it before the meeting tomorrow

 I don't have time to do profiling today

> I meant profiling the current situation (master) so that you have a feeling at which places how much time is lost and plan the optimizations based on real data.

The new code is completely different, profiling `master` first is of very little use. A benchmark run to compare performance may be interesting, but that can be done afterwards thanks to git.

> kdbGet is also the operation where it is more important that it is fast.

I agree. But you commented on a step of `kdbSet`.

> so that we see that it is not much slower after the PR.

Even if it is slower after the PR, we should merge this. The PR was started because of a major new feature. If the first implementation is slow, we can always improve it later. Better we get a version out there, so people get use to the new contracts, specifications and conventions.

Reply:
Discussion result: Any optimization that is already implement and working will be kept in place. No new optimizations will be added. At the end of the PR we will evaluate the existing optimizations and remove any that don't help with performance.